rfcs/converged-implementation-patterns.md

```diff
@@ -0,0 +1,258 @@
+# RFC: Converged component implementation pattern
+
+<!--
+An RFC can be anything. A question, a suggestion, a plan. The purpose of this template is to give some structure to help folks write successful RFCs. However, don't feel constrained by this template; use your best judgement.
+
+Tips for writing a successful RFC:
+
+- Simple plain words that make your point, fancy words obfuscate
+- Try to stay concise, but don't gloss over important details
+- Try to write a neutral problem statement, not one that motivates your desired solution
+- Remember, "Writing is thinking". It's natural to realize new ideas while writing your proposal
+-->
+
+---
+
+_List contributors to the proposal here_
+
+## Summary
+
+<!-- Explain the proposed change -->
+
+Tries to commit the common component pattern to writing in the repo for future reference and track improvements
+
+Code samples should applied to utility scripts to generate packages which will help to educate/enforce this pattern... until we want to change it :-)
+
+## Background
+
+<!-- If there is relevant background include it here -->
+
+Many new engineers can find it hard to adapt to a new codebase, especially when the coding patterns and practices are fairly new and in a constant state of flux
+
+## Problem statement
+
+<!--
+Why are we making this change? What problem are we solving? What do we expect to gain from this?
+
+This section is important as the motivation or problem statement is indepenent from the proposed change. Even if this RFC is not accepted this Motivation can be used for alternative solutions.
+
+In the end, please make sure to present a neutral Problem statement, rather than one that motivates a particular solution
+-->
+
+Many new developers might have questions on the patterns that we follow to write a React component. This RFC attempts to commit and possibly improve our code patterns in writing so that there might be a reference for new (and old) members of the team.
+
+## Detailed Design or Proposal
+
+<!-- This is the bulk of the RFC. Explain the proposal or design in enough detail for the inteded audience to understand. -->
+
+### Hook based architecture.
+
+Generally a component will have the following different files. Let us consider a `Sample` component
+
+- Sample.tsx
+- useSample.ts
+- useSampleClasses/useSampleStyles.ts
+- renderSample.tsx
+- Sample.types.ts
+
+#### Sample.tsx
+
+The 'final' product, simply forwards a ref and uses all the previous building blocks to expose the final component
```

Comment:
```suggestion
The 'final' product, simply forwards a ref and uses all the previous building blocks to expose the final component. Also connects styling.
```

?

Reply:
Yeah, that does make sense, but what if we add more hooks ? I was thinking of `useSampleAccessibility`

The point is that this file should base itself primarily off the tools in the module

Reply:
Q: how does this connects styling ? If I don't call `useSampleStyles()` explicitly, no changes to styling would be applied.

Reply:
well that's the point right ? by calling `useSampleStyles` you are styling the component

Reply:
yes, but there is no explicit/enforced connection, that's what I had in mind :) 

_Sorry for another out of topic again:_ 
Also this api enables user to mutating state, which may not be the best solution from my experience, especially in terms of VDOM rendering. Is this "a feature" that we wanna encourage our consumers to go with?